plugins/com.google.cloud.tools.eclipse.preferences.test/src/com/google/cloud/tools/eclipse/preferences/areas/AreaPreferencePageTest.java

```diff
@@ -0,0 +1,240 @@
+
+@RunWith(MockitoJUnitRunner.class)
+public class AreaPreferencePageTest {
+  private @Mock IPersistentPreferenceStore preferences;
+
+  private Shell shell;
+
+  @After
+  public void tearDown() {
+    if (shell != null && !shell.isDisposed()) {
+      shell.dispose();
+    }
+  }
+
+  @Test
+  public void testWritesOnApply() {
+    // add an Area, call performApply, verify 1 call to preferences.store()
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    page.addArea(new TestPrefArea("pref1", "value", preferences));
+    show(page);
+    assertTrue(page.performOk());
+    verify(preferences).setValue("pref1", "value");
+  }
+
+  @Test
+  public void testNoWritesOnCancel() {
+    // add an Area, call performCancel, verify no calls to preferences.store()
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    page.addArea(new TestPrefArea("pref1", "value", preferences));
+    show(page);
+
+    page.performCancel();
+    verify(preferences, never()).setValue(anyString(), anyString());
+  }
+
+  @Test
+  public void testErrorStatusResults() {
+    // add an Area, trigger its VALUE property, cause its status to be error.
+    // verify showing error message
+    AreaBasedPreferencePage page = new AreaBasedPreferencePage("test");
+    TestPrefArea area = new TestPrefArea("pref1", "value", preferences);
+    page.addArea(area);
+
+    show(page);
+    assertTrue(page.isValid());
+    assertEquals(IMessageProvider.NONE, page.getMessageType());
+
+    area.status = new Status(IStatus.ERROR, "foo", "message");
+    area.fireValueChanged(TestPrefArea.IS_VALID, true, false);
+
+    assertFalse(page.isValid());
+    assertEquals(IMessageProvider.ERROR, page.getMessageType());
+    assertEquals("message", page.getMessage());
+
+    // change its status to be OK, trigger its VALUE property.
+    // verify no error message
+    area.status = Status.OK_STATUS;
+    area.fireValueChanged(TestPrefArea.IS_VALID, false, true);
+
+    assertTrue(page.isValid());
+    assertEquals(IMessageProvider.NONE, page.getMessageType());
+  }
+
+  private TestPrefArea area1;
+  private TestPrefArea area2;
+  private TestPrefArea area3;
+  private AreaBasedPreferencePage page;
+
+  private void setupAreas() {
+    page = new AreaBasedPreferencePage("test");
+    area1 = new TestPrefArea("pref1", "value", preferences);
+    area2 = new TestPrefArea("pref2", "value", preferences);
+    area3 = new TestPrefArea("pref3", "value", preferences);
+    page.addArea(area1);
+    page.addArea(area2);
+    page.addArea(area3);
+
+    area1.status = Status.OK_STATUS;
+    area2.status = Status.OK_STATUS;
+    area3.status = Status.OK_STATUS;
+
+    show(page);
+    assertTrue(page.isValid());
+    assertEquals(IMessageProvider.NONE, page.getMessageType());
+  }
+
+  @Test
+  public void testStatusSeverityInfo() {
+    // add three areas, status = OK, OK, INFO
+    // verify showing INFO
+    // verify can apply
```

Comment:
I removed it as (1) the apply button is hidden and has to be explicitly exposed, and (2) the apply button's enablement is actually controlled by the result of `isValid()`, which is changed by calls to `setValid()`.  So by checking isValid() we implicitly know the apply state.